source/server/debug_session_properties_restricted_service.cpp

```diff
@@ -0,0 +1,244 @@
+  NISysCfgEnumResourceHandle resources_handle = NULL;
+  NISysCfgResourceHandle resource = NULL;
+
+  try {
+    auto library = get_syscfg_library_interface();
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library->CreateFilter(session, &filter))) {
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyIsDevice, NISysCfgBoolTrue);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyProductId, product_id);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertySerialNumber, serial_number);
+        if (NISysCfg_Succeeded(status = library->FindHardware(session, NISysCfgFilterModeMatchValuesAll, filter, NULL, &resources_handle))) {
+          if (NISysCfg_Succeeded(status) && (status = library->NextResource(session, resources_handle, &resource)) == NISysCfg_OK) {
+            library->GetResourceProperty(resource, property_id, value);
+            library->CloseHandle(resource);
+          }
+          library->CloseHandle(resources_handle);
+        }
+        library->CloseHandle(filter);
+      }
+    }
+  }
+  catch (nidevice_grpc::NonDriverException& ex) {
+    return ex.GetStatus();
+  }
+
+  if (NISysCfg_Failed(status)) {
+    return ::grpc::Status(::grpc::StatusCode::INTERNAL, kDebugSessionPropertyAccessFailedMessage);
+  }
+
+  return ::grpc::Status::OK;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::set_bool_property(
+  NISysCfgResourceProperty property_id,
+  uint32_t product_id,
+  const char* serial_number,
+  bool value)
+{
+  NISysCfgStatus status = NISysCfg_OK;
+  NISysCfgSessionHandle session = NULL;
+  NISysCfgFilterHandle filter = NULL;
+  NISysCfgEnumResourceHandle resources_handle = NULL;
+  NISysCfgResourceHandle resource = NULL;
+  NISysCfgBool syscfg_value = value ? NISysCfgBoolTrue : NISysCfgBoolFalse;
+
+  try {
+    auto library = get_syscfg_library_interface();
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library->CreateFilter(session, &filter))) {
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyIsDevice, NISysCfgBoolTrue);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyProductId, product_id);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertySerialNumber, serial_number);
+        if (NISysCfg_Succeeded(status = library->FindHardware(session, NISysCfgFilterModeMatchValuesAll, filter, NULL, &resources_handle))) {
+          if (NISysCfg_Succeeded(status) && (status = library->NextResource(session, resources_handle, &resource)) == NISysCfg_OK) {
+            library->SetResourceProperty(resource, property_id, syscfg_value);
+            NISysCfgBool changes_require_restart = NISysCfgBoolFalse;
+            char* detailed_changes = nullptr;
+            library->SaveResourceChanges(resource, &changes_require_restart, &detailed_changes);
```

Comment:
You need to check the status here and return it if saving fails. Also, on failure, `detailed_changes` should be returned as the error message.

Reply:
Assigned status to the result of SaveResourceChanges().  I did not retain the detailed_changes for the error message due to the reasons I outlined in the above conversation.